generic ephemeral volume E2E tests

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Some more tests for the feature are useful.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
storage e2e: test snapshotting of generic ephemeral volumes

Conceptually, snapshots have to be taken while the pod and thus the volume
exist. Snapshotting has an issue where flushing of data is not guaranteed while
the volume is still staged on the node, so the test relied on deleting the pod
and checking for the volume to be unused. That part of the test cannot be done
for ephmeral volumes.

pohly committed:
storage e2e: test volume attach limits of generic ephemeral volumes

There are unit tests for this particular code path in kube-scheduler, but no
E2E tests.

pohly committed:
storage e2e: fix volume metric test for generic ephemeral volume

This is a fix for the new test case from
https://github.com/kubernetes/kubernetes/pull/105636 which had to be merged
without prior testing due to not having a cluster to test on and no pull job
which runs these
tests. https://testgrid.k8s.io/sig-storage-kubernetes#gce-serial then showed a
failure.

The fix is simple: in the ephemeral case, the PVC name isn't set in advance in
pvc.Name and instead must be computed. The fix now was tested on a kubetest
cluster in GCE.

Signed-off-by: Patrick Ohly <patrick.ohly@intel.com>

mauriciopoppe commented:
this is in my TODO list, list of items that I'll read:

- The KEP
- The [testplan](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/1698-generic-ephemeral-volumes/README.md#test-plan) in the KEP
- https://github.com/kubernetes/kubernetes/pull/92784
- The attached PRs

pohly commented:
/retest

mauriciopoppe reviewed (commented):
Left one open question for the volumesnapshot test

pohly committed:
storage e2e: refactor snapshottable

During PR review it was pointed out that the branches for ephemeral
vs. persistent make the test harder to read. Therefore all code that depends on
if checks gets moved into two different versions of the test, one hat runs for
ephemeral volumes and one for persistent volumes, with skip statements at the
beginning.

mauriciopoppe commented:
/retest

mauriciopoppe commented:
/lgtm
